shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

```diff
@@ -179,6 +257,48 @@ public SpannableStringBuilder replace(
       Log.e(TAG, "editing state should not be changed in a listener callback");
     }
 
+    Log.e(
+        "DELTAS",
+        "replace(" + start + ", " + end + ", " + tb + ", " + tbstart + ", " + tbend + ")");
+
+    final boolean isDeletionGreaterThanOne = end - (start + tbend) > 1;
+    final boolean isCalledFromDelete = tb == "" && tbstart == 0 && tbstart == tbend;
+
+    final boolean isReplacedByShorter = isDeletionGreaterThanOne && (tbend - tbstart < end - start);
+    final boolean isReplacedByLonger = tbend - tbstart > end - start;
+    final boolean isReplacedBySame = tbend - tbstart == end - start;
+
+    // Is deleting/inserting at the end of a composing region.
+    final boolean isDeletingInsideComposingRegion = !isReplacedByShorter && start + tbend < end;
+    final boolean isInsertingInsideComposingRegion = start + tbend > end;
+
+    // To consider the cases when autocorrect increases the length of the text being composed by
+    // one, but changes more than one character.
+    final boolean isOriginalComposingRegionTextChanged =
+        (isCalledFromDelete || isDeletingInsideComposingRegion || isReplacedByShorter)
+            || !toString()
+                .subSequence(start, end)
+                .equals(tb.toString().subSequence(tbstart, end - start));
+
+    // A replacement means the original composing region has changed, anything else will be
+    // considered an insertion.
+    final boolean isReplaced =
+        isOriginalComposingRegionTextChanged
+            && (isReplacedByLonger || isReplacedBySame || isReplacedByShorter);
+
+    if (isCalledFromDelete || isDeletingInsideComposingRegion) {
+      Log.e("DELTAS", "DELETION");
+      setDeltas(
+          toString(), toString().subSequence(start + tbend, end).toString(), "DELETION", end, end);
+    } else if ((start == end || isInsertingInsideComposingRegion)
+        && !isOriginalComposingRegionTextChanged) {
+      Log.e("DELTAS", "INSERTION");
+      setDeltas(toString(), tb.subSequence(end - start, tbend).toString(), "INSERTION", end, end);
+    } else if (isReplaced) {
+      Log.e("DELTAS", "REPLACEMENT");
+      setDeltas(toString(), tb.subSequence(tbstart, tbend).toString(), "REPLACEMENT", start, end);
+    }
+
```

Comment:
This brings me back to thinking that it's possible to infer the type of TextEditingDelta without explicitly saving it, as I commented about in the design doc: https://docs.google.com/document/d/1PXNd_LwJudiUTnCM2AXRxbd98_p4BlFf6qFhEhv3wvs/edit?pli=1&disco=AAAAN4YZXnA

Can you give me an example of ambiguity that requires a type?

Reply:
If we do need types, maybe consider making the strings (e.g. "REPLACEMENT") into constants.